### ISSUE TITLE
how to manually update mongo journal from 0.4 to 1.0?

### ISSUE THREAD

**Issue by alari** [scullxbones/akka-persistence-mongo#44]
Hi.

I'm trying to migrate to Akka 2.4 and plugin 1.0.0, but auto-migrator doesn't work as expected.

How can I manually migrate journal to the new version?

**Comment by scullxbones** [comment]
Sorry to hear that - would definitely like to hear more about it.  Are you seeing an error on upgrade?  Does nothing happen?  Does the journal get corrupted?  If I could somehow get a copy of your journal that would help - although I understand if it can't be shared.

The new data model has two major changes - it is batched (1 or more events per document) and there is a type hint.  

Here's a sample:

```
{
  "pid": "akka",
  "from": 1,
  "to": 10,
  "v": 1,
  "events":[{
    "pid": "akka",
    "sn": 1,
    "manifest":"sample test",
    "_t":"bson",
    "p":{
      "free":"form",
      "bson":"payload"
    },
   "v": 1
  }, ..., {
    "pid": "akka",
    "sn": 10,
    "_t":"ser",
    "p":BinData("..."),
    "_h":"com.foo.bar.OpenEvent", // Type hint
    "v":1
  }]
}
```

So it may be possible to run a js to upgrade the data - I may need to do something about the fact that the plugin unpacks the payload from received `PersistentRepr`s (see the type hint in the ser type).  If you're using only bson types (the PR you contributed :smile:), then it should be very doable in a js script.

For the rest, it would be very useful to understand how the automated upgrade is failing.  That's why any more details you have about what you're seeing would help me out a lot.

**Comment by alari** [comment]
Hi.

I'm using only bson types (except for sharding). After I compiled the project with 1.0.0 and changed configs to launch auto-upgrade, nothing happened. Journal was not changed.

I can provide you with journal dump and any logs tomorrow, hope it will help.

Thanks!

**Comment by scullxbones** [comment]
Ok, let me check some things out. I have a passing spec for the upgrade directly, but maybe the config part has a stupid mistake in it.

**Comment by scullxbones** [comment]
Well it seems to work assuming the following is used:

```akka.contrib.persistence.mongodb.mongo.journal-automatic-upgrade = true```

I'm interested in seeing exactly what's going on for sure.

**Comment by alari** [comment]
Is there any way to debug journal-automatic-update?

Plugin's code looks like it just should work. I use config option just as it's stated in the code. But there's no logging during journal updating. Have it found objects? Is it launching at all? What's happening?

I'm running it with Play and akka sharding, maybe something interfere.

**Comment by scullxbones** [comment]
That's a good idea - I can add some logging in and put out a point release.  It wouldn't be very obtrusive since it should only run once for a journal and it must be enabled.  I'm going to change this ticket to "add some logging"

**Comment by scullxbones** [comment]
So I added a bunch of logging around release `1.0.1` - please give it a shot and let's see if we can get to the bottom of this.

Thanks!

**Comment by alari** [comment]
```
[error] a.c.p.m.RxMongoDriver - Upgrade did not complete successfully
java.lang.IllegalStateException: Unable to deserialize PersistentRepr
	at akka.contrib.persistence.mongodb.RxMongoSerializers$JournalDeserializer$$anonfun$1.apply(RxMongoSerializers.scala:129) ~[akka-persistence-mongo-rxmongo_2.11-1.0.1.jar:1.0.1]
	at akka.contrib.persistence.mongodb.RxMongoSerializers$JournalDeserializer$$anonfun$1.apply(RxMongoSerializers.scala:129) ~[akka-persistence-mongo-rxmongo_2.11-1.0.1.jar:1.0.1]
	at scala.util.Try.getOrElse(Try.scala:79) ~[scala-library-2.11.7.jar:na]
	at akka.contrib.persistence.mongodb.RxMongoSerializers$JournalDeserializer$.deserializeDocumentLegacy(RxMongoSerializers.scala:129) ~[akka-persistence-mongo-rxmongo_2.11-1.0.1.jar:1.0.1]
	at akka.contrib.persistence.mongodb.RxMongoSerializers$JournalDeserializer$.deserializeDocument(RxMongoSerializers.scala:88) ~[akka-persistence-mongo-rxmongo_2.11-1.0.1.jar:1.0.1]
	at akka.contrib.persistence.mongodb.RxMongoSerializers$JournalDeserializer$.deserializeDocument(RxMongoSerializers.scala:84) ~[akka-persistence-mongo-rxmongo_2.11-1.0.1.jar:1.0.1]
	at akka.contrib.persistence.mongodb.MongoPersistenceDriver.deserializeJournal(MongoPersistence.scala:107) ~[akka-persistence-mongo-common_2.11-1.0.1.jar:1.0.1]
	at akka.contrib.persistence.mongodb.RxMongoDriver.walk(RxMongoPersistenceExtension.scala:77) ~[akka-persistence-mongo-rxmongo_2.11-1.0.1.jar:1.0.1]
	at akka.contrib.persistence.mongodb.RxMongoDriver$$anonfun$2.apply(RxMongoPersistenceExtension.scala:93) ~[akka-persistence-mongo-rxmongo_2.11-1.0.1.jar:1.0.1]
	at akka.contrib.persistence.mongodb.RxMongoDriver$$anonfun$2.apply(RxMongoPersistenceExtension.scala:93) ~[akka-persistence-mongo-rxmongo_2.11-1.0.1.jar:1.0.1]
[ERROR] [09/07/2015 11:52:22.167] [application-akka.actor.default-dispatcher-15] [akka.tcp://application@127.0.0.1:60620/system/sharding/orders/0/5382f68a00702513f0d758de] Persistence failure when replaying events for persistenceId [order-5382f68a00702513f0d758de]. Last known sequence number [0]
java.lang.IllegalStateException: Unable to deserialize PersistentRepr
	at akka.contrib.persistence.mongodb.RxMongoSerializers$JournalDeserializer$$anonfun$1.apply(RxMongoSerializers.scala:129)
	at akka.contrib.persistence.mongodb.RxMongoSerializers$JournalDeserializer$$anonfun$1.apply(RxMongoSerializers.scala:129)
	at scala.util.Try.getOrElse(Try.scala:79)
	at akka.contrib.persistence.mongodb.RxMongoSerializers$JournalDeserializer$.deserializeDocumentLegacy(RxMongoSerializers.scala:129)
	at akka.contrib.persistence.mongodb.RxMongoSerializers$JournalDeserializer$.deserializeDocument(RxMongoSerializers.scala:88)
	at akka.contrib.persistence.mongodb.RxMongoSerializers$JournalDeserializer$.deserializeDocument(RxMongoSerializers.scala:84)
	at akka.contrib.persistence.mongodb.MongoPersistenceDriver.deserializeJournal(MongoPersistence.scala:107)
	at akka.contrib.persistence.mongodb.RxMongoDriver.walk(RxMongoPersistenceExtension.scala:77)
	at akka.contrib.persistence.mongodb.RxMongoDriver$$anonfun$2.apply(RxMongoPersistenceExtension.scala:93)
	at akka.contrib.persistence.mongodb.RxMongoDriver$$anonfun$2.apply(RxMongoPersistenceExtension.scala:93)
	at reactivemongo.api.DefaultCursor$$anon$2.reactivemongo$api$DefaultCursor$$anon$$go$1(cursor.scala:332)
	at reactivemongo.api.DefaultCursor$$anon$2$$anonfun$foldWhile$2.apply(cursor.scala:335)
	at reactivemongo.api.DefaultCursor$$anon$2$$anonfun$foldWhile$2.apply(cursor.scala:335)
	at reactivemongo.api.DefaultCursor$$anon$2$$anonfun$foldBulks$2.apply(cursor.scala:313)
	at reactivemongo.api.DefaultCursor$$anon$2$$anonfun$foldBulks$2.apply(cursor.scala:311)
	at reactivemongo.api.DefaultCursor$$anon$2$FoldResponses$$anonfun$procResponses$1.apply(cursor.scala:455)
	at reactivemongo.api.DefaultCursor$$anon$2$FoldResponses$$anonfun$procResponses$1.apply(cursor.scala:454)
	at scala.concurrent.Future$$anonfun$flatMap$1.apply(Future.scala:251)
	at scala.concurrent.Future$$anonfun$flatMap$1.apply(Future.scala:249)
	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:32)
	at scala.concurrent.impl.ExecutionContextImpl$AdaptedForkJoinTask.exec(ExecutionContextImpl.scala:121)
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.pollAndExecAll(ForkJoinPool.java:1253)
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1346)
	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

**Comment by alari** [comment]
````json
{
	"_id" : ObjectId("55deeae33de20e69f33b748b"),
	"pid" : "order-5382f68a00702513f0d758de",
	"sn" : NumberLong(1),
	"dl" : false,
	"cs" : [ ],
	"pr" : {
		"p" : {
			"order-created" : {
				"id" : "5382f68a00702513f0d758de",
				"seqNr" : 156,
				"userId" : "5382f68900702513f0d758dd",
				"cartId" : "???",
				"phoneNumber" : "49114888198",
				"from" : {
					"country" : "US"
				},
				"to" : {
					"country" : "RU",
					"region" : "MOW",
					"city" : "Moscow"
				},
				"dateCreated" : ISODate("2015-08-27T10:48:03.101Z"),
				"timestamp" : ISODate("2015-08-27T10:48:03.101Z"),
				"addressId" : "7ec53cbb-93fe-4eff-8f98-7d5e1846151f"
			},
			"_timestamp" : ISODate("2015-08-27T10:48:03.102Z")
		}
	}
}
{
	"_id" : ObjectId("55deeae33de20e69f33b764e"),
	"pid" : "order-5382f68a00702513f0d758de",
	"sn" : NumberLong(2),
	"dl" : false,
	"cs" : [ ],
	"pr" : {
		"p" : {
			"state-forced" : {
				"state" : "Invalid",
				"timestamp" : ISODate("2015-08-27T10:48:03.392Z")
			},
			"_timestamp" : ISODate("2015-08-27T10:48:03.392Z")
		}
	}
}
{
	"_id" : ObjectId("55deec183de20e69f33b7728"),
	"pid" : "order-5382f68a00702513f0d758de",
	"sn" : NumberLong(3),
	"dl" : false,
	"cs" : [ ],
	"pr" : {
		"p" : {
			"order-created" : {
				"id" : "5382f68a00702513f0d758de",
				"seqNr" : 526,
				"userId" : "5382f68900702513f0d758dd",
				"cartId" : "???",
				"phoneNumber" : "49114888198",
				"from" : {
					"country" : "US"
				},
				"to" : {
					"country" : "RU",
					"region" : "MOW",
					"city" : "Moscow"
				},
				"dateCreated" : ISODate("2015-08-27T10:53:12.764Z"),
				"timestamp" : ISODate("2015-08-27T10:53:12.764Z"),
				"addressId" : "7ec53cbb-93fe-4eff-8f98-7d5e1846151f"
			},
			"_timestamp" : ISODate("2015-08-27T10:53:12.765Z")
		}
	}
}
{
	"_id" : ObjectId("55deec193de20e69f33b78d6"),
	"pid" : "order-5382f68a00702513f0d758de",
	"sn" : NumberLong(4),
	"dl" : false,
	"cs" : [ ],
	"pr" : {
		"p" : {
			"state-forced" : {
				"state" : "Invalid",
				"timestamp" : ISODate("2015-08-27T10:53:13.104Z")
			},
			"_timestamp" : ISODate("2015-08-27T10:53:13.104Z")
		}
	}
}
```

**Comment by alari** [comment]
so it fails with the first journal message

**Comment by scullxbones** [comment]
Thanks for the repeatable case!  I should be able to quickly figure out what's up.

**Comment by scullxbones** [comment]
Hmm.  So that passes the test - it seems to upgrade correctly.

The error message suggests that the driver thinks the payload is binary, so i'm wondering if somehow the logging is reporting the wrong record.  I added some more information to the exception message in version `1.0.2`.  Let me know what else you see.

**Comment by alari** [comment]
The reason was in another message. Now logging is much better.

I removed all sharding messages, and now having the following error:

```
reactivemongo.api.commands.UpdateWriteResult: DatabaseException['E11000 duplicate key error index: gipsetter.persistent.journal.$journal.index dup key: { : "cart-c39ef9d3-06ab-4065-8bda-e46fc2eca0f1", : null, : null }' (code = 11000)]
```

I think old journal indexes should be dropped before upgrade.

**Comment by alari** [comment]
With extended logging I was able to perform migration -- with some usage of console.

Thanks!

**Comment by scullxbones** [comment]
That's great!  Is there anything that you did in the console I can roll into the upgrade code?

> I removed all sharding messages, and now having the following error

I have a feeling between this and #41 that cluster sharding journal entries are *not* backward compatible 2.3 to 2.4.  I probably need to loop in the akka team to at least take a look.  What error were you getting with these?  Or were these the unserializable `PersistentRepr`s?

> I think old journal indexes should be dropped before upgrade.

I thought about this, but I didn't want to drop them until the user was happy with the upgrade.  Given that field names changed, they shouldn't cause a problem - it looks like you did run into something there though.  Ah, yes I see that it's treating non-existent `SequenceNr` and `Deleted` as null.  Ok, so that I can automate.

Please let me know if anything else comes up with the 2.4 code.

**Comment by alari** [comment]
Sharding messages are not binary compatible between contrib (2.3) and stable (2.4) versions.

I'm doing two things in console:

- Dropping index.
- Removing all sharding messages by pattern.

`mongo gipsetter --eval "db.persistent.journal.remove({pid:{'$regex':'^/user/sharding/'}});db.persistent.journal.dropIndex('journal.index’)”`

Both points are usable for everyone and could be automated.

**Comment by scullxbones** [comment]
I added the `dropIndex` to the journal upgrade function in version `1.0.3`.  I'm hesitant to automatically delete any `persistentId`s that look like cluster sharding.  I want to make sure that the akka team knows about that incompatibility - they may want to do things another way or document the issue.

For users of this library I plan on putting your remove step above in the `akka24.md` readme, should they want to address the problem.

Thanks for the help!

**Comment by scullxbones** [comment]
After touching base with @patriknw, he has agreed that older cluster sharding entries should be removed.  I'll also add that to the upgrade process.

**Comment by patriknw** [comment]
https://github.com/akka/akka/pull/18436

**Comment by scullxbones** [comment]
Closed (for the last time?  :smile:) with `1.0.4`

**Comment by alari** [comment]
Great, thank you!
